openedx/core/djangoapps/content/learning_sequences/api/processors/special_exams.py

```diff
@@ -0,0 +1,81 @@
+import logging
+from datetime import datetime
+
+import six
+from edx_proctoring.api import get_attempt_status_summary
+from edx_proctoring.exceptions import ProctoredExamNotFoundException
+from django.conf import settings
+from django.contrib.auth import get_user_model
+from opaque_keys.edx.keys import CourseKey
+from util import milestones_helpers
+
+from ...data import SpecialExamAttemptData, UserCourseOutlineData
+from .base import OutlineProcessor
+
+User = get_user_model()
+log = logging.getLogger(__name__)
+
+
+class SpecialExamsOutlineProcessor(OutlineProcessor):
+    """
+    Responsible for applying all outline processing related to special exams.
+    """
+    def load_data(self):
+        """
+        Check if special exams are enabled
+        """
+        self.special_exams_enabled = settings.FEATURES.get('ENABLE_SPECIAL_EXAMS', False)
+
+    def usage_keys_to_remove(self, full_course_outline):
```

Comment:
Here we only need to remove the xblocks from the outline if `not self.include_special_exams is True`. But since we always include special exams (unlike in the transformers), why do we need this? Maybe I am looking at the wrong place?
https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_api/blocks/transformers/milestones.py#L77

Reply:
If you look a [few lines up in the old transformer](https://github.com/edx/edx-platform/blob/master/lms/djangoapps/course_api/blocks/transformers/milestones.py#L75), you'll see that check is also dependent on the value of `settings.FEATURES.ENABLE_SPECIAL_EXAMS`. In this case, we are gating in a similar way

Reply:
The way I am reading it is if `ENABLE_SPECIAL_EXAMS is True` we will NOT gate the block and if `ENABLE_SPECIAL_EXAMS is False` then to gate it `include_special_exams` must be `False` which it never is here. So the value of `ENABLE_SPECIAL_EXAMS` should not impact whether we gate it or not. The only thing this processor should do is add the user exam attempts data.

Reply:
That's not quite accurate. This method is `usage_keys_to_remove`.

If `ENABLE_SPECIAL_EXAMS` is `False`, all exams will be completely removed from the outline (not gated). If `ENABLE_SPECIAL_EXAMS` is True, then all exams will remain as is in the outline.

With regards to the `exam_data` method.

If `ENABLE_SPECIAL_EXAMS` is `False`, no attempts will be made for any sequences to fetch exam data. If `ENABLE_SPECIAL_EXAMS` is True, then all exam sequences will have attempt data fetched.

Reply:
Oh okay, got it.

Reply:
@symbolist, that's correct. I removed this.